distributed/shuffle/_rechunk.py

```diff
@@ -360,93 +364,87 @@ def __init__(
         self.worker_for = worker_for
         self.split_axes = split_axes(old, new)
 
-    async def _receive(self, data: list[tuple[NDIndex, bytes]]) -> None:
+    async def _receive(
+        self,
+        data: list[tuple[NDIndex, list[tuple[NDIndex, tuple[NDIndex, np.ndarray]]]]],
+    ) -> None:
         self.raise_if_closed()
 
-        filtered = []
+        # Repartition shards and filter out already received ones
+        shards = defaultdict(list)
         for d in data:
-            id, payload = d
-            if id in self.received:
+            id1, payload = d
+            if id1 in self.received:
                 continue
-            filtered.append(payload)
-            self.received.add(id)
+            self.received.add(id1)
+            for id2, shard in payload:
+                shards[id2].append(shard)
             self.total_recvd += sizeof(d)
         del data
-        if not filtered:
+        if not shards:
             return
+
         try:
-            shards = await self.offload(self._repartition_shards, filtered)
-            del filtered
             await self._write_to_disk(shards)
         except Exception as e:
             self._exception = e
             raise
 
-    def _repartition_shards(self, data: list[bytes]) -> dict[NDIndex, bytes]:
-        repartitioned: defaultdict[
-            NDIndex, list[tuple[NDIndex, np.ndarray]]
-        ] = defaultdict(list)
-        for buffer in data:
-            for id, shard in pickle.loads(buffer):
-                repartitioned[id].append(shard)
-        return {k: pickle.dumps(v) for k, v in repartitioned.items()}
-
     async def _add_partition(
         self, data: np.ndarray, partition_id: NDIndex, **kwargs: Any
     ) -> int:
-        def _() -> dict[str, tuple[NDIndex, bytes]]:
-            """Return a mapping of worker addresses to a tuple of input partition
-            IDs and shard data.
-
-
-            TODO: Overhaul!
-            As shard data, we serialize the payload together with the sub-index of the
-            slice within the new chunk. To assemble the new chunk from its shards, it
-            needs the sub-index to know where each shard belongs within the chunk.
-            Adding the sub-index into the serialized payload on the sender allows us to
-            write the serialized payload directly to disk on the receiver.
-            """
-            out: dict[
-                str, list[tuple[NDIndex, tuple[NDIndex, np.ndarray]]]
-            ] = defaultdict(list)
-            from itertools import product
-
-            ndsplits = product(
-                *(axis[i] for axis, i in zip(self.split_axes, partition_id))
+        out = defaultdict(list)
```

Comment:
that may be a hotfix like I was referring to in https://github.com/dask/distributed/pull/8282#issuecomment-1778735818 but that likely requires a little testing.

Reply:
This is very interesting, I'll play around with it.

Reply:
It seems that it's already so. The input of `_send` is already a `_List`.

Reply:
IIUC this means that the shards are pickled / copied before it is sent because it basically calls `pickle.dumps(_List(...))` instead of doing something smart. However, in some cases this is desired since iterating through the list can be expensive.

Reply:
@fjetter to my understanding it's calling pickle.dumps with buffer_callback. which is exactly what we want.

Reply:
I may have misstated my earlier statement a little. What `_List` is doing is it prohibits us recursing into nested collections. This is implemented here https://github.com/dask/distributed/blob/b62b70051eee77684e7964f88a6553aeb453bc2b/distributed/protocol/serialize.py#L303-L325

It's sometimes hard to tell how that translates to real world examples and therefore I wrote this down with examples https://gist.github.com/fjetter/8e7a06b664838b37433f889c6ce63ca5#file-nestedcollectionsserialization-ipynb

We're using this kind of nesting when defining the shards in a couple of places. For instance, the dataframe shuffling is defining the shards as tuples where the first item is an integer while the second part is a arrow table. This will inevitably cause a full pickle+data copy
https://github.com/dask/distributed/blob/b62b70051eee77684e7964f88a6553aeb453bc2b/distributed/shuffle/_shuffle.py#L500

However, we're more or less intentionally casting the list of shards to `_List` because iterating through the list of shards has been shown to be very, very slow in the past. This may no longer be the case, just something to be aware of. I encourage folks to remove the `_List` again and try out what happens.